processing/src/main/java/org/apache/druid/segment/column/CapabilitiesBasedFormat.java

```diff
@@ -111,18 +110,18 @@ public ColumnFormat merge(@Nullable ColumnFormat otherFormat)
     } else if (!Objects.equals(merged.getComplexTypeName(), otherSnapshot.getComplexTypeName())) {
       throw new ISE(
           "Cannot merge columns of type[%s] and [%s]",
-          merged.getComplexTypeName(),
-          otherSnapshot.getComplexTypeName()
+          mergedType,
+          otherType
       );
     }
 
     merged.setDictionaryEncoded(merged.isDictionaryEncoded().or(otherSnapshot.isDictionaryEncoded()).isTrue());
     merged.setHasMultipleValues(merged.hasMultipleValues().or(otherSnapshot.hasMultipleValues()).isTrue());
     merged.setDictionaryValuesSorted(
-        merged.areDictionaryValuesSorted().and(otherSnapshot.areDictionaryValuesSorted()).isTrue()
+        merged.areDictionaryValuesSorted().or(otherSnapshot.areDictionaryValuesSorted()).isTrue()
```

Comment:
Doesn't the correct choice depend on how the actual segment merging is expected to happen? Like:

- if the merging preserves sortedness as long as both inputs are sorted, it should be `and`
- if the merging always sorts regardless of input sortedness, it should just be hard coded to `TRUE`
- if the merging sorts if _either_ input is sorted, it should be `or`
- if the merging requires that both inputs have the same sortedness, it should be an error if the two inputs have different sortedness

this change should be in a separate PR anyway, since it's a behavioral change beyond improving the exception message.

Reply:
To add, your comment highlights why `ColumnFormat` exists in the first place, which is because simply merging `ColumnCapabilities` in a generic/universal manner isn't really powerful enough to model these things in a satisfying way.

Reply:
TY for the commentary, that makes sense.